### PR TITLE
[scheduler] Add build+test scheduling

### PR DESCRIPTION
**Pull request by CaseyHillers**
This is a draft to estimate the work needed if we wanted to implement a build+test scheduler.

- [x] Postsubmit processor
- [x] Presubmit processor
- [x] Nice to have: Refactor the common code into a library

Fixes https://github.com/flutter/flutter/issues/119036

### CODE REVIEW

**Comment by CaseyHillers** [comment]
\cc @yjbanov. I don't there's much complexity to add this, but it would be nice if whoever authors this can refactor the common code into a library.

File: app_dart/lib/src/request_handlers/scheduler/postsubmit_build_test_subscription.dart [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+import '../../model/luci/push_message.dart';
+import '../../request_handling/body.dart';
+import '../../request_handling/exceptions.dart';
+import '../../request_handling/subscription_handler.dart';
+import '../../service/datastore.dart';
+import '../../service/logging.dart';
+import '../../service/scheduler.dart';
+
+// TODO(chillers): Create new subscription.
+// TODO(chillers): Write tests.
```
Comment (ricardoamador) [inline app_dart/lib/src/request_handlers/scheduler/postsubmit_build_test_subscription.dart:20]: Is there a reason tests cannot be added now?

File: app_dart/lib/src/request_handlers/scheduler/presubmit_build_test_subscription.dart [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    final String builderName = build.tagsByName('builder').first;
+    log.fine('Available tags: ${build.tags.toString()}');
+    final List<String> tags = build.tagsByName('buildset');
+    final String? prNumberRaw = tags.firstWhereOrNull((String tag) => tag.contains('pr/git/'));
+    if (prNumberRaw == null) {
+      log.warning('Build does not contain buildset tag with pr/git/####');
+      return Body.empty;
+    }
+    final int prNumber = int.parse(prNumberRaw.replaceAll('pr/git/', ''));
+    final String sha = tags.singleWhere((String tag) => tag.contains('sha/git/')).replaceAll('sha/git/', '');
```
Comment (ricardoamador) [inline app_dart/lib/src/request_handlers/scheduler/presubmit_build_test_subscription.dart:51]: Make constants for 'sha/git/' and probably 'pr/git' if used in multiple places.

**Comment by godofredoc** [comment]
Is there a design doc for this?

**Comment by CaseyHillers** [comment]
> Is there a design doc for this?

There's go/flutter-build-test-infra, but I acknowledge it lacks implementation details. I can send a one pager next week covering the plan here with migration guide + customer's asking for this feature.

**Comment by CaseyHillers** [comment]
As an update, I wrote a draft of a design doc at https://docs.google.com/document/d/1TC3YaFnYK_9jtaRnoZofHphYBsRNIt-YFmwgcetCDzM/edit. We're waiting on @keyonghan to return to review this, and align it with the rest of the infra build-test plan.

**Comment by CaseyHillers** [comment]
Closing. @yjbanov mentioned there was a solution for this that unblocked the web team. @keyonghan this could be useful in the future to optimize our infra fleet, and prevent orchestrator bots from being needed.

**Pull request closed without merging**